File: Lab(ICE-2202)CPP/lab-04.cpp
#include <iostream>
#include <vector>
using namespace std;

int binary_search(vector<int> &arr, int target)
{
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
        {
            return mid;
        }
        else if (arr[mid] < target)
        {
            left = mid + 1;
        }
        else
        {
            right = mid - 1;
        }
    }

    return -1; // Return -1 if element not found
}

int main()
{
    int n;
    cout << "Enter the size of the array: ";
    cin >> n;

    vector<int> arr(n);
    cout << "Enter the sorted elements of the array: ";
    for (int i = 0; i < n; ++i)
    {
        cin >> arr[i];
    }

    int target;
    cout << "Enter the element to search for: ";
    cin >> target;

    int index = binary_search(arr, target);

    if (index != -1)
    {
        cout << "Element " << target << " found at index " << index << endl;
    }
    else
    {
        cout << "Element " << target << " not found in the array." << endl;
    }

    return 0;
}

File: lab-04.py
def binary_search(arr, target):
    """
    Perform binary search to find the target element in the sorted array.
    """
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1


# Input section to take a sorted list and the target element from the user
arr = list(map(int, input("Enter a sorted list of numbers separated by space: ").split()))
target = int(input("Enter the element to search for: "))

# Perform binary search
index = binary_search(arr, target)

# Output the result
if index != -1:
    print("Element", target, "found at index:", index)
else:
    print("Element", target, "not found in the list.")
